L1Trigger/TrackFindingTMTT/src/Histos.cc

```diff
@@ -0,0 +1,1776 @@
+#include "L1Trigger/TrackFindingTMTT/interface/Histos.h"
+#include "L1Trigger/TrackFindingTMTT/interface/InputData.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Sector.h"
+#include "L1Trigger/TrackFindingTMTT/interface/HTrphi.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Make3Dtracks.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TrkRZfilter.h"
+#include "L1Trigger/TrackFindingTMTT/interface/L1fittedTrack.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Utility.h"
+#include "L1Trigger/TrackFindingTMTT/interface/PrintL1trk.h"
+
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "DataFormats/Math/interface/deltaR.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "TH1F.h"
+#include "TH2F.h"
+#include "TH2Poly.h"
+#include "TF1.h"
+#include "TPad.h"
+#include "TProfile.h"
+#include "TGraphAsymmErrors.h"
+#include "TGraph.h"
+#include "TEfficiency.h"
+
+#include <algorithm>
+#include <array>
+#include <unordered_set>
+#include <list>
+#include <sstream>
+#include <memory>
+#include <mutex>
+
+using namespace std;
+
+namespace tmtt {
+
+  namespace {
+    std::once_flag printOnce;
+    std::mutex myMutex[6];
```

Comment:
Is it the intended behaviour that two of the functions protected by the different mutex may be called concurrently ?
If the different functions should be mutually exclusive, this should be a single `mutex` object, and used by all functions.
If the different functions **can** the used concurrently, I would suggest moving the various `mutex` objects to be `static` variables inside the respective functions.

Reply:
The Histos class fill histograms within edm::stream::EDProducer. We need to avoid that two threads attempt to fill the same histogram at the same time. This could be achieved with a single mutex lock covering all functions that fill histograms. However, if we did this, whilst one stream was working its way through these functions, the other streams would have to wait. Instead therefore, we use an independent mutex lock for each function, so that one stream can be filling the histograms related to one function, whilst a second stream can be filling the histograms related to another function.